### ISSUE TITLE
Support for Ubisys C4

### ISSUE THREAD

**Issue by theFork** [Koenkk/zigbee-herdsman-converters#973]
I want to add support for the quad push-button controller C4 by Ubisys. This is what I found out so far:
```
{
        zigbeeModel: ['C4 (5504)'],
        model: 'C4',
        vendor: 'Ubisys',
        description: 'Quad button C4',
        supports: 'four push buttons',

        fromZigbee: [???],
}
```
The device gets successfully paired to zigbee2mqtt, but does not seem to send any messages. As far as I remember from working with another coordinator (Dresden Raspbee / Deconz) I needed to set a group for each of the buttons in order to make them work. Most probably we will need to configure the device before it sends button push events.

A detailled tech reference document can be found here:
https://www.ubisys.de/wp-content/uploads/ubisys-c4-technical-reference-1.pdf

Unfortunetely, my Zigbee knowledge ends right here. But I can test all suggestions you make.

**Comment by sjorge** [comment]
Maybe @felixstorm can help, IIRC he did the support for the other Ubisys devices.

**Comment by felixstorm** [comment]
Actually I do have this on my list, but the list is long :wink: and AFAIR implementing support for ubisys C4 is not a simple task. IIRC the C4 will not send anything out to the coordinator without first having been paired to another ZigBee device - and I am not even sure if pairing to the coordinator would work (but it probably will). And to configure the C4 to do more than the most basic stuff, we will even need to use ZigBee commands that have not yet been implemented in zigbee-herdsman (we need Write Attributes *Structured* as writing the complete attribute data in one go will lead to packets too large for ZigBee).

Long story short: Yes, I would like to implement support for the ubisys C4 some time, but it will probably be at least a few weeks from now until I'll find the time...

@theFork What would be your use case? Just receiving key presses within `zigbee2mqtt`? Or binding to other ZigBee devices (and if so - which devices)?

**Comment by theFork** [comment]
Hi Felix, thanks for your reply, I would be thrilled if you find the time :) Yes, this is what I also observed, the C4 does not send anything when not configured.

My use cases are:
* Receive key presses via MQTT (i.e. trigger an automation that starts webradio)
* Bind to Zigbee-capable lightbulbs (I can test a Hue, several Ikea Tradfri and an Innr device)
* Bind to one channel of a Ubisys S2 switch

**Comment by theFork** [comment]
One additional note:
I assume that the C4 will broadcast messages in case a we assign a (dummy-) group to each button. This is what helped me back when I was using deconz. As far as I understand Zigbee, the group approach would also allow to bind devices by adding them to the group.

**Comment by felixstorm** [comment]
@theFork working on it (https://github.com/felixstorm/zigbee-shepherd-converters/commit/ab25788ab12781784af17b40b9a9a4aaff0d1b9b) - binding to the coordinator does seem to work. will hopefully be able to have initial support ready within the next 3 days or 1 to 2 weeks, need to work on the InputActions attribute now to make buttons configurable...

**Comment by felixstorm** [comment]
first PRs:  
* https://github.com/Koenkk/zigbee-herdsman-converters/pull/1013
* https://github.com/Koenkk/zigbee-herdsman/pull/142

**Comment by felixstorm** [comment]
Final PR created to configure inputs for the C4, so this issue should then hopefully be solved after the PR has been merged and released...

**Comment by Koenkk** [comment]
PRs have been merged so I assume this can be closed now. Thanks!

**Comment by theFork** [comment]
Thanks for your effort! Just for the sake of completeness, I have tested some of the implemented features and would like to document what I have tested so far:

**1. Unbound**
We get these log messages, when hitting an unconfigured/unbound button of a paired C4:
```
zigbee2mqtt:info  2020-03-03 18:16:59: MQTT publish: topic 'zigbee2mqtt/eg_flur_c4', payload '{"linkquality":57,"action":"2_toggle"}'
zigbee2mqtt:info  2020-03-03 18:16:59: MQTT publish: topic 'zigbee2mqtt/eg_flur_c4', payload '{"linkquality":57,"action":""}'
```
Looks good to me.

**2. Binding**
As a second test I tried binding the fourth button of a C4 to an IKEA Tradfri bulb:
```
command: zigbee2mqtt/bridge/bind/0x001fee00000039c6/4
payload: 0x000b57fffe99aa9f
```
Resulting in the following log:

```
zigbee2mqtt:info  2020-03-01 18:52:17: Successfully bound cluster 'genScenes' from 'eg_flur_c4' to 'eg_flur_stimmungslicht'
zigbee2mqtt:info  2020-03-01 18:52:17: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"device_bind","message":{"from":"eg_flur_c4","to":"eg_flur_stimmungslicht","cluster":"genScenes"}}'
zigbee2mqtt:info  2020-03-01 18:52:18: Successfully bound cluster 'genOnOff' from 'eg_flur_c4' to 'eg_flur_stimmungslicht'
zigbee2mqtt:info  2020-03-01 18:52:18: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"device_bind","message":{"from":"eg_flur_c4","to":"eg_flur_stimmungslicht","cluster":"genOnOff"}}'
zigbee2mqtt:info  2020-03-01 18:52:19: Successfully bound cluster 'genLevelCtrl' from 'eg_flur_c4' to 'eg_flur_stimmungslicht'
zigbee2mqtt:info  2020-03-01 18:52:19: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"type":"device_bind","message":{"from":"eg_flur_c4","to":"eg_flur_stimmungslicht","cluster":"genLevelCtrl"}}'
```
I'm a bit curious, why my command initiated binding to 3 different clusters. I would have expected only the OnOff cluster. What are the endpoints that I can bind? A little bit documentation on that would be appreciated.

Nevertheless, the button switches the bulb as I requested. Sometimes it appears, as if the bulb gets toggled a second time, about 3 seconds after I pressed the switch. Maybe this is due to the Tradfri bulb, I will have to investigate on this a little further and open another issue if necessary.

Again, thanks so much, this really made Z2M usable for me!

**Comment by felixstorm** [comment]
@theFork Binding binds all matching clusters (from a list, currently `['genScenes', 'genOnOff', 'genLevelCtrl', 'lightingColorCtrl', 'closuresWindowCovering']`) between the devices. Since all three you mentioned are on the list of *output* clusters of the C4 and on the list of *input* clusters on the bulb, z2m binds all three of them.

To be able to use scenes or level control from the C4, you would need to configure the C4 inputs to actually do something different than sending toggle commands for the on/off cluster (as it does in it's factory reset configuration). Unfortunately the deep down technical details and examples on how to do this are - as of now - available from ubisys only upon request (called "Input Configuration Integrator's Guide"). But if you take a look at the publicly available technical reference guide for the device you want to control (take the D1 guide for standard light bulbs), you will find their native configuration (look for the attribute `InputActions`) which can be adapted for the C4 also (you mainly need to change the endpoint to point to the correct one for the C4).  

I am currently in conversation with ubisys about some technical details and also wanted to ask them to make the integrator's guide available publicly, and hope to get some answer back from them this week or next. In case they do not want to publish the integrator's guide, I will probably at least post some standard configurations on the [C4 device page](https://github.com/Koenkk/zigbee2mqtt.io/blob/develop/docs/devices/C4.md) so that people can use the standard ones without having to contact ubisys.

**Comment by felixstorm** [comment]
If you want to bind your C4 to a light bulb and use on/off and dimming with **one** input/momentary switch/push button, you could use the following configuration:
![image](https://user-images.githubusercontent.com/1177810/75857543-ac2a1b00-5df6-11ea-8cec-21dcbbc39602.png)
Change the input number (0 in the example) to the one you want to use and also the endpoint (although **2** which is used in the example above should also work fine on the C4 - you just need to remember to bind this endpoint to the bulb thereafter), convert the numbers to decimal and everything into a JSON array of arrays (no need for the length, has been stroked out in the example above) and publish it as `configure_device_setup` to `zigbee2mqtt/[FRIENDLY_NAME]/set`. Then your C4 should be able to control a bound bulb very neatly.

Another example (this time already in JSON format - used it for testing myself): Control a (hue) bulb with the first **two** buttons from endpoint **1**:
```json
{
    "configure_device_setup": {
        "inputActions": [
            [ 0,  7,  1,  6,  0,  1 ],
            [ 0,  6,  1,  8,  0,  5,  0, 50 ],
            [ 0, 11,  1,  8,  0,  7 ],
            [ 1,  7,  1,  6,  0,  0 ],
            [ 1,  6,  1,  8,  0,  5,  1, 50 ],
            [ 1, 11,  1,  8,  0,  7 ]
        ]
    }
}
```

In case you want to reset your C4 to it's factory default configuration this is also available on the [C4 device page](https://github.com/Koenkk/zigbee2mqtt.io/blob/develop/docs/devices/C4.md#configuring-inputs).

**Comment by felixstorm** [comment]
Some more information on the C4 endpoints: There are 4 available with on/off, level and scenes (1-4, for lights) and 2 for window coverings (5-6).  
You only need to use the number of endpoints corresponding to the number of different client devices you want to control, e.g. if you control two lights from input 1 and 2 and one window cover from inputs 3 (move up) and 4 (move down), you would use and bind endpoints 1 and 2 to your bulbs and endpoint 5 to your window cover. Then you need to configure your inputs to act on the corresponding endpoints and you should be good to go.
